.builders/tests/test_upload.py

```diff
@@ -275,11 +274,102 @@ def test_upload_built_existing_different_sha_does_upload_multiple_existing_build
         'existing-1.1.1-cp311-cp311-manylinux2010_x86_64.whl': new_hash,
     })
 
-    upload.upload(targets_dir)
+    targets = upload.upload(targets_dir)
 
     uploads = {str(Path(f).name) for f in uploads}
 
     assert uploads == {'existing-1.1.1-cp311-cp311-manylinux2010_x86_64.whl'}
 
     bucket_files = {f.name for f in bucket.list_blobs()}
     assert f'built/existing/existing-1.1.1-{frozen_timestamp}-cp311-cp311-manylinux2010_x86_64.whl' in bucket_files
+
+
+# Test when there are built and external packages for a dependency and the external wheel is the prefered dependency
+def test_external_wheel_priority(tmp_path, setup_targets_dir, setup_fake_bucket, setup_fake_hash):
+    original_hash = 'first-hash'
+    external_hash = 'external-hash'
+
+    # in this pipeline run, we made an external wheel for the 'existing' dependency
+    wheels = {
+        'external': [
+            ('existing-1.1.1-cp312-cp312-manylinux2010_x86_64.whl', 'existing', '1.1.1', '>=3.7'),
+        ]
+    }
+
+    targets_dir = setup_targets_dir(wheels)
+
+    bucket_files = {
```

Comment:
I just want to confirm that these version numbers are all synthetic test data and not referring to the pinned versions we really use. Othewise this will brittle over updates.

Reply:
Correct!